diff --git a/src/ggml-cuda/fattn-tile.cu b/src/ggml-cuda/fattn-tile.cu
index e563729a..93b1b00b 100644
--- a/src/ggml-cuda/fattn-tile.cu
+++ b/src/ggml-cuda/fattn-tile.cu
@@ -5,6 +5,18 @@ void ggml_cuda_flash_attn_ext_tile(ggml_backend_cuda_context & ctx, ggml_tensor
     const ggml_tensor * K = dst->src[1];
     const ggml_tensor * V = dst->src[2];
     switch (K->ne[0]) {
+        case  16: {
+            GGML_ASSERT(V->ne[0] == K->ne[0]);
+            ggml_cuda_flash_attn_ext_tile_case< 16,  16>(ctx, dst);
+        } break;
+        case  32: {
+            GGML_ASSERT(V->ne[0] == K->ne[0]);
+            ggml_cuda_flash_attn_ext_tile_case< 32,  32>(ctx, dst);
+        } break;
+        case  56: {
+            GGML_ASSERT(V->ne[0] == K->ne[0]);
+            ggml_cuda_flash_attn_ext_tile_case< 56,  56>(ctx, dst);
+        } break;
         case  40: {
             GGML_ASSERT(V->ne[0] == K->ne[0]);
             ggml_cuda_flash_attn_ext_tile_case< 40,  40>(ctx, dst);
diff --git a/src/ggml-cuda/fattn-tile.cuh b/src/ggml-cuda/fattn-tile.cuh
index d1164b85..d7d7f73a 100644
--- a/src/ggml-cuda/fattn-tile.cuh
+++ b/src/ggml-cuda/fattn-tile.cuh
@@ -19,6 +19,23 @@
     }                                                                                                 \
 
 static constexpr __host__ __device__ uint32_t ggml_cuda_fattn_tile_get_config_nvidia_fp16(const int DKQ, const int DV, const int ncols) {
+    GGML_CUDA_FATTN_TILE_CONFIG_CASE( 16,  16,  2,  64, 2,  32,  16)
+    GGML_CUDA_FATTN_TILE_CONFIG_CASE( 16,  16,  4, 128, 2,  32,  16)
+    GGML_CUDA_FATTN_TILE_CONFIG_CASE( 16,  16,  8, 256, 2,  32,  16)
+    GGML_CUDA_FATTN_TILE_CONFIG_CASE( 16,  16, 16, 256, 2,  32,  16)
+    GGML_CUDA_FATTN_TILE_CONFIG_CASE( 16,  16, 32, 256, 2,  32,  16)
+
+    GGML_CUDA_FATTN_TILE_CONFIG_CASE( 32,  32,  2,  64, 2,  32,  32)
+    GGML_CUDA_FATTN_TILE_CONFIG_CASE( 32,  32,  4, 128, 2,  32,  32)
+    GGML_CUDA_FATTN_TILE_CONFIG_CASE( 32,  32,  8, 256, 2,  32,  32)
+    GGML_CUDA_FATTN_TILE_CONFIG_CASE( 32,  32, 16, 256, 2,  32,  32)
+    GGML_CUDA_FATTN_TILE_CONFIG_CASE( 32,  32, 32, 256, 2,  32,  32)
+
+    GGML_CUDA_FATTN_TILE_CONFIG_CASE( 56,  56,  2,  64, 2,  64,  56)
+    GGML_CUDA_FATTN_TILE_CONFIG_CASE( 56,  56,  4, 128, 2,  64,  56)
+    GGML_CUDA_FATTN_TILE_CONFIG_CASE( 56,  56,  8, 256, 2,  64,  56)
+    GGML_CUDA_FATTN_TILE_CONFIG_CASE( 56,  56, 16, 256, 2,  64,  56)
+    GGML_CUDA_FATTN_TILE_CONFIG_CASE( 56,  56, 32, 256, 2,  64,  56)
     GGML_CUDA_FATTN_TILE_CONFIG_CASE( 40,  40,  2,  64, 2,  64,  40)
     GGML_CUDA_FATTN_TILE_CONFIG_CASE( 40,  40,  4, 128, 2,  64,  40)
     GGML_CUDA_FATTN_TILE_CONFIG_CASE( 40,  40,  8, 256, 2,  64,  40)
@@ -88,6 +105,23 @@ static constexpr __host__ __device__ uint32_t ggml_cuda_fattn_tile_get_config_nv
 }
 
 static constexpr __host__ __device__ uint32_t ggml_cuda_fattn_tile_get_config_nvidia_fp32(const int DKQ, const int DV, const int ncols) {
+    GGML_CUDA_FATTN_TILE_CONFIG_CASE( 16,  16,  2,  64, 2,  32,  16)
+    GGML_CUDA_FATTN_TILE_CONFIG_CASE( 16,  16,  4, 128, 2,  32,  16)
+    GGML_CUDA_FATTN_TILE_CONFIG_CASE( 16,  16,  8, 256, 2,  32,  16)
+    GGML_CUDA_FATTN_TILE_CONFIG_CASE( 16,  16, 16, 256, 2,  32,  16)
+    GGML_CUDA_FATTN_TILE_CONFIG_CASE( 16,  16, 32, 256, 2,  32,  16)
+
+    GGML_CUDA_FATTN_TILE_CONFIG_CASE( 32,  32,  2,  64, 2,  32,  32)
+    GGML_CUDA_FATTN_TILE_CONFIG_CASE( 32,  32,  4, 128, 2,  32,  32)
+    GGML_CUDA_FATTN_TILE_CONFIG_CASE( 32,  32,  8, 256, 2,  32,  32)
+    GGML_CUDA_FATTN_TILE_CONFIG_CASE( 32,  32, 16, 256, 2,  32,  32)
+    GGML_CUDA_FATTN_TILE_CONFIG_CASE( 32,  32, 32, 256, 2,  32,  32)
+
+    GGML_CUDA_FATTN_TILE_CONFIG_CASE( 56,  56,  2,  64, 2,  64,  56)
+    GGML_CUDA_FATTN_TILE_CONFIG_CASE( 56,  56,  4, 128, 2,  64,  56)
+    GGML_CUDA_FATTN_TILE_CONFIG_CASE( 56,  56,  8, 256, 2,  64,  56)
+    GGML_CUDA_FATTN_TILE_CONFIG_CASE( 56,  56, 16, 256, 2,  64,  56)
+    GGML_CUDA_FATTN_TILE_CONFIG_CASE( 56,  56, 32, 256, 2,  64,  56)
     GGML_CUDA_FATTN_TILE_CONFIG_CASE( 40,  40,  2,  64, 2,  32,  40)
     GGML_CUDA_FATTN_TILE_CONFIG_CASE( 40,  40,  4, 128, 2,  32,  40)
     GGML_CUDA_FATTN_TILE_CONFIG_CASE( 40,  40,  8, 256, 2,  32,  40)
@@ -157,6 +191,26 @@ static constexpr __host__ __device__ uint32_t ggml_cuda_fattn_tile_get_config_nv
 }
 
 static constexpr __host__ __device__ uint32_t ggml_cuda_fattn_tile_get_config_amd(const int DKQ, const int DV, const int ncols) {
+    GGML_CUDA_FATTN_TILE_CONFIG_CASE( 16,  16,  2,  64, 2,  32,  16)
+    GGML_CUDA_FATTN_TILE_CONFIG_CASE( 16,  16,  4, 128, 2,  32,  16)
+    GGML_CUDA_FATTN_TILE_CONFIG_CASE( 16,  16,  8, 256, 2,  32,  16)
+    GGML_CUDA_FATTN_TILE_CONFIG_CASE( 16,  16, 16, 256, 2,  32,  16)
+    GGML_CUDA_FATTN_TILE_CONFIG_CASE( 16,  16, 32, 256, 2,  32,  16)
+    GGML_CUDA_FATTN_TILE_CONFIG_CASE( 16,  16, 64, 256, 2,  32,  16)
+
+    GGML_CUDA_FATTN_TILE_CONFIG_CASE( 32,  32,  2,  64, 2,  32,  32)
+    GGML_CUDA_FATTN_TILE_CONFIG_CASE( 32,  32,  4, 128, 2,  32,  32)
+    GGML_CUDA_FATTN_TILE_CONFIG_CASE( 32,  32,  8, 256, 2,  32,  32)
+    GGML_CUDA_FATTN_TILE_CONFIG_CASE( 32,  32, 16, 256, 2,  32,  32)
+    GGML_CUDA_FATTN_TILE_CONFIG_CASE( 32,  32, 32, 256, 2,  32,  32)
+    GGML_CUDA_FATTN_TILE_CONFIG_CASE( 32,  32, 64, 256, 2,  32,  32)
+
+    GGML_CUDA_FATTN_TILE_CONFIG_CASE( 56,  56,  2,  64, 2,  64,  56)
+    GGML_CUDA_FATTN_TILE_CONFIG_CASE( 56,  56,  4, 128, 2,  64,  56)
+    GGML_CUDA_FATTN_TILE_CONFIG_CASE( 56,  56,  8, 256, 2,  64,  56)
+    GGML_CUDA_FATTN_TILE_CONFIG_CASE( 56,  56, 16, 256, 2,  64,  56)
+    GGML_CUDA_FATTN_TILE_CONFIG_CASE( 56,  56, 32, 256, 2,  64,  56)
+    GGML_CUDA_FATTN_TILE_CONFIG_CASE( 56,  56, 64, 256, 2,  64,  56)
     GGML_CUDA_FATTN_TILE_CONFIG_CASE( 40,  40,  2,  64, 2,  32,  40)
     GGML_CUDA_FATTN_TILE_CONFIG_CASE( 40,  40,  4, 128, 2,  32,  40)
     GGML_CUDA_FATTN_TILE_CONFIG_CASE( 40,  40,  8, 256, 2,  32,  40)
@@ -235,6 +289,26 @@ static constexpr __host__ __device__ uint32_t ggml_cuda_fattn_tile_get_config_am
 }
 
 static constexpr __host__ __device__ uint32_t ggml_cuda_fattn_tile_get_config_amd_rdna(const int DKQ, const int DV, const int ncols) {
+    GGML_CUDA_FATTN_TILE_CONFIG_CASE( 16,  16,  2,  64, 2,  32,  16)
+    GGML_CUDA_FATTN_TILE_CONFIG_CASE( 16,  16,  4, 128, 2,  32,  16)
+    GGML_CUDA_FATTN_TILE_CONFIG_CASE( 16,  16,  8, 256, 2,  32,  16)
+    GGML_CUDA_FATTN_TILE_CONFIG_CASE( 16,  16, 16, 256, 2,  32,  16)
+    GGML_CUDA_FATTN_TILE_CONFIG_CASE( 16,  16, 32, 256, 2,  32,  16)
+    GGML_CUDA_FATTN_TILE_CONFIG_CASE( 16,  16, 64, 256, 2,  32,  16)
+
+    GGML_CUDA_FATTN_TILE_CONFIG_CASE( 32,  32,  2,  64, 2,  32,  32)
+    GGML_CUDA_FATTN_TILE_CONFIG_CASE( 32,  32,  4, 128, 2,  32,  32)
+    GGML_CUDA_FATTN_TILE_CONFIG_CASE( 32,  32,  8, 256, 2,  32,  32)
+    GGML_CUDA_FATTN_TILE_CONFIG_CASE( 32,  32, 16, 256, 2,  32,  32)
+    GGML_CUDA_FATTN_TILE_CONFIG_CASE( 32,  32, 32, 256, 2,  32,  32)
+    GGML_CUDA_FATTN_TILE_CONFIG_CASE( 32,  32, 64, 256, 2,  32,  32)
+
+    GGML_CUDA_FATTN_TILE_CONFIG_CASE( 56,  56,  2,  64, 2,  64,  56)
+    GGML_CUDA_FATTN_TILE_CONFIG_CASE( 56,  56,  4, 128, 2,  64,  56)
+    GGML_CUDA_FATTN_TILE_CONFIG_CASE( 56,  56,  8, 256, 2,  64,  56)
+    GGML_CUDA_FATTN_TILE_CONFIG_CASE( 56,  56, 16, 256, 2,  64,  56)
+    GGML_CUDA_FATTN_TILE_CONFIG_CASE( 56,  56, 32, 256, 2,  64,  56)
+    GGML_CUDA_FATTN_TILE_CONFIG_CASE( 56,  56, 64, 256, 2,  64,  56)
     GGML_CUDA_FATTN_TILE_CONFIG_CASE( 40,  40,  2,  64, 2,  32,  40)
     GGML_CUDA_FATTN_TILE_CONFIG_CASE( 40,  40,  4, 128, 2,  32,  40)
     GGML_CUDA_FATTN_TILE_CONFIG_CASE( 40,  40,  8, 256, 2,  32,  40)
@@ -1341,6 +1415,9 @@ void ggml_cuda_flash_attn_ext_tile(ggml_backend_cuda_context & ctx, ggml_tensor
     template void ggml_cuda_flash_attn_ext_tile_case              \
     <DKQ, DV>(ggml_backend_cuda_context & ctx, ggml_tensor * dst) \
 
+extern DECL_FATTN_TILE_CASE( 16,  16);
+extern DECL_FATTN_TILE_CASE( 32,  32);
+extern DECL_FATTN_TILE_CASE( 56,  56);
 extern DECL_FATTN_TILE_CASE( 40,  40);
 extern DECL_FATTN_TILE_CASE( 64,  64);
 extern DECL_FATTN_TILE_CASE( 72,  72);
diff --git a/src/ggml-cuda/fattn.cu b/src/ggml-cuda/fattn.cu
index ab7a3b29..ad83ecad 100644
--- a/src/ggml-cuda/fattn.cu
+++ b/src/ggml-cuda/fattn.cu
@@ -391,7 +391,10 @@ static best_fattn_kernel ggml_cuda_get_best_fattn_kernel(const int device, const
     const int cc = ggml_cuda_info().devices[device].cc;
 
     switch (K->ne[0]) {
+        case  16:
+        case  32:
         case  40:
+        case  56:
         case  64:
         case  72:
         case  80:
@@ -458,7 +461,7 @@ static best_fattn_kernel ggml_cuda_get_best_fattn_kernel(const int device, const
     const bool can_use_vector_kernel = Q->ne[0] <= 256 && Q->ne[0] % 64 == 0 && Q->ne[0] != 192 && K->ne[1] % FATTN_KQ_STRIDE == 0;
 
     // If Turing tensor cores are available, use them:
-    if (turing_mma_available(cc) && Q->ne[0] != 40 && Q->ne[0] != 72) {
+    if (turing_mma_available(cc) && Q->ne[0] != 16 && Q->ne[0] != 32 && Q->ne[0] != 40 && Q->ne[0] != 56 && Q->ne[0] != 72) {
         if (can_use_vector_kernel) {
             if (!ggml_is_quantized(K->type) && !ggml_is_quantized(V->type)) {
                 if (cc >= GGML_CUDA_CC_ADA_LOVELACE && Q->ne[1] == 1 && Q->ne[3] == 1 && !(gqa_ratio > 4 && K->ne[1] >= 8192)) {
@@ -488,7 +491,7 @@ static best_fattn_kernel ggml_cuda_get_best_fattn_kernel(const int device, const
         gqa_ratio_eff *= 2;
     }
 
-    if (volta_mma_available(cc) && Q->ne[0] != 40 && Q->ne[0] != 72) {
+    if (volta_mma_available(cc) && Q->ne[0] != 16 && Q->ne[0] != 32 && Q->ne[0] != 40 && Q->ne[0] != 56 && Q->ne[0] != 72) {
         if (can_use_vector_kernel && Q->ne[1] * gqa_ratio_eff <= 2) {
             return BEST_FATTN_KERNEL_VEC;
         }
@@ -499,7 +502,7 @@ static best_fattn_kernel ggml_cuda_get_best_fattn_kernel(const int device, const
     }
 
     // AMD MFMA needs a certain minimum batch size to outscale the tile kernel for large head sizes.
-    if ((amd_mfma_available(cc) && Q->ne[0] <= 256) && Q->ne[0] != 40 && Q->ne[0] != 72) {
+    if ((amd_mfma_available(cc) && Q->ne[0] <= 256) && Q->ne[0] != 16 && Q->ne[0] != 32 && Q->ne[0] != 40 && Q->ne[0] != 56 && Q->ne[0] != 72) {
         if ((Q->ne[0] <= 64 && Q->ne[1] * gqa_ratio_eff > 8)) {
             return BEST_FATTN_KERNEL_MMA_F16;
         }
@@ -512,7 +515,7 @@ static best_fattn_kernel ggml_cuda_get_best_fattn_kernel(const int device, const
     }
 
     // AMD WMMA is always faster than the tile kernel if the full tile width of 16 can be utilized.
-    if ((amd_wmma_available(cc) && gqa_opt_applies && Q->ne[0] <= 128) && Q->ne[0] != 40 && Q->ne[0] != 72 && Q->ne[1] * gqa_ratio_eff > 8) {
+    if ((amd_wmma_available(cc) && gqa_opt_applies && Q->ne[0] <= 128) && Q->ne[0] != 16 && Q->ne[0] != 32 && Q->ne[0] != 40 && Q->ne[0] != 56 && Q->ne[0] != 72 && Q->ne[1] * gqa_ratio_eff > 8) {
         return BEST_FATTN_KERNEL_MMA_F16;
     }
 
diff --git a/src/ggml-cuda/template-instances/fattn-tile-instance-dkq16-dv16.cu b/src/ggml-cuda/template-instances/fattn-tile-instance-dkq16-dv16.cu
new file mode 100644
index 00000000..088b27f7
--- /dev/null
+++ b/src/ggml-cuda/template-instances/fattn-tile-instance-dkq16-dv16.cu
@@ -0,0 +1,5 @@
+// This file has been autogenerated by generate_cu_files.py, do not edit manually.
+
+#include "../fattn-tile.cuh"
+
+DECL_FATTN_TILE_CASE(16, 16);
diff --git a/src/ggml-cuda/template-instances/fattn-tile-instance-dkq32-dv32.cu b/src/ggml-cuda/template-instances/fattn-tile-instance-dkq32-dv32.cu
new file mode 100644
index 00000000..45bc7fcd
--- /dev/null
+++ b/src/ggml-cuda/template-instances/fattn-tile-instance-dkq32-dv32.cu
@@ -0,0 +1,5 @@
+// This file has been autogenerated by generate_cu_files.py, do not edit manually.
+
+#include "../fattn-tile.cuh"
+
+DECL_FATTN_TILE_CASE(32, 32);
diff --git a/src/ggml-cuda/template-instances/fattn-tile-instance-dkq56-dv56.cu b/src/ggml-cuda/template-instances/fattn-tile-instance-dkq56-dv56.cu
new file mode 100644
index 00000000..00edd305
--- /dev/null
+++ b/src/ggml-cuda/template-instances/fattn-tile-instance-dkq56-dv56.cu
@@ -0,0 +1,5 @@
+// This file has been autogenerated by generate_cu_files.py, do not edit manually.
+
+#include "../fattn-tile.cuh"
+
+DECL_FATTN_TILE_CASE(56, 56);
